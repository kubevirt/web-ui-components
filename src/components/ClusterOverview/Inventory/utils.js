import { getVmStatus, VM_STATUS_ALL_ERROR, VM_STATUS_ALL_PROGRESS } from '../../../utils/status/vm';
import { getPodStatus, POD_STATUS_ALL_ERROR, POD_STATUS_ALL_PROGRESS } from '../../../utils/status/pod';
import { getNodeStatus, NODE_STATUS_ALL_ERROR, NODE_STATUS_ALL_WARN } from '../../../utils/status/node';
import { getPvcStatus, PVC_STATUS_ALL_ERROR, PVC_STATUS_ALL_PROGRESS } from '../../../utils/status/pvc';

// same as InventoryItemStatus props
const STATUS_RESULT_OK = 'ok';
const STATUS_RESULT_WARN = 'warn';
const STATUS_RESULT_ERROR = 'error';
const STATUS_RESULT_IN_PROGRESS = 'inProgress';

const mapStatuses = (entities, mapEntityToStatusResult) => {
  const result = {
    [STATUS_RESULT_OK]: 0,
    [STATUS_RESULT_WARN]: 0,
    [STATUS_RESULT_ERROR]: 0,
    [STATUS_RESULT_IN_PROGRESS]: 0,
    count: null,
  };

  if (entities) {
    entities.forEach(entity => {
      const resolvedStatusResult = (mapEntityToStatusResult && mapEntityToStatusResult(entity)) || STATUS_RESULT_OK;
      result[resolvedStatusResult]++;
    });
    result.count = entities.length;
  }

  return result;
};

const resolveStatusResult = (status, mapper) =>
  Object.keys(mapper).find(key => mapper[key].includes(status)) || STATUS_RESULT_OK;

export const mapNodesToProps = nodes =>
  mapStatuses(nodes, node =>
    resolveStatusResult(getNodeStatus(node).status, {
      [STATUS_RESULT_ERROR]: NODE_STATUS_ALL_ERROR,
      [STATUS_RESULT_WARN]: NODE_STATUS_ALL_WARN,
    })
  );

export const mapVmsToProps = (vms, pods, migrations) =>
  mapStatuses(vms, vm =>
    resolveStatusResult(getVmStatus(vm, pods, migrations).status, {
      [STATUS_RESULT_ERROR]: VM_STATUS_ALL_ERROR,
      [STATUS_RESULT_IN_PROGRESS]: VM_STATUS_ALL_PROGRESS,
    })
  );

export const mapPvcsToProps = pvcs =>
  mapStatuses(pvcs, pvc =>
    resolveStatusResult(getPvcStatus(pvc).status, {
      [STATUS_RESULT_ERROR]: PVC_STATUS_ALL_ERROR,
      [STATUS_RESULT_IN_PROGRESS]: PVC_STATUS_ALL_PROGRESS,
    })
  );

export const mapPodsToProps = pods =>
  mapStatuses(pods, pod =>
    resolveStatusResult(getPodStatus(pod).status, {
      [STATUS_RESULT_ERROR]: POD_STATUS_ALL_ERROR,
      [STATUS_RESULT_IN_PROGRESS]: POD_STATUS_ALL_PROGRESS,
    })
  );
