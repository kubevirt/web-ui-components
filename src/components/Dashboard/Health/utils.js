import { get } from 'lodash';

import { LOADING_STATE, OK_STATE, ERROR_STATE, WARNING_STATE } from './HealthItem';
import {
  OCS_HEALTHY,
  OCS_DEGRADED,
  OCS_ERROR,
  OCS_UNKNOWN,
  OCP_HEALTHY,
  OCP_ERROR,
  CNV_HEALTHY,
  CNV_ERROR,
} from './strings';

export const getKubevirtHealthState = kubevirtHealth => {
  if (!kubevirtHealth) {
    return { state: LOADING_STATE };
  }
  return get(kubevirtHealth, 'apiserver.connectivity') === 'ok'
    ? { message: CNV_HEALTHY, state: OK_STATE }
    : { message: CNV_ERROR, state: ERROR_STATE };
};

export const getK8sHealthState = k8sHealth => {
  if (!k8sHealth) {
    return { state: LOADING_STATE };
  }
  return get(k8sHealth, 'response') === 'ok'
    ? { message: OCP_HEALTHY, state: OK_STATE }
    : { message: OCP_ERROR, state: ERROR_STATE };
};

const OCSHealthStatus = {
  0: {
    message: OCS_HEALTHY,
    state: OK_STATE,
  },
  1: {
    message: OCS_DEGRADED,
    state: WARNING_STATE,
  },
  2: {
    message: OCS_ERROR,
    state: ERROR_STATE,
  },
  3: {
    message: OCS_UNKNOWN,
    state: ERROR_STATE,
  },
};

export const getOCSHealthState = ocsResponse => {
  if (!ocsResponse) {
    return { state: LOADING_STATE };
  }
  const value = get(ocsResponse, 'result[0].value[1]');
  return OCSHealthStatus[value] || OCSHealthStatus[3];
};
