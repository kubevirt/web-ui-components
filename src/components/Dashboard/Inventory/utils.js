import {
  getVmStatus,
  VM_STATUS_ALL_WARNING,
  VM_STATUS_ALL_ERROR,
  VM_STATUS_ALL_PROGRESS,
  VM_STATUS_ALL_OFF,
} from '../../../utils/status/vm';
import { getPodStatus, POD_STATUS_ALL_ERROR, POD_STATUS_ALL_PROGRESS } from '../../../utils/status/pod';
import { getNodeStatus, NODE_STATUS_ALL_ERROR, NODE_STATUS_ALL_WARN } from '../../../utils/status/node';
import { getPvcStatus, PVC_STATUS_ALL_ERROR, PVC_STATUS_ALL_PROGRESS } from '../../../utils/status/pvc';
import {
  getHostStatus,
  HOST_STATUS_ALL_ERROR,
  HOST_STATUS_ALL_WARN,
  HOST_STATUS_ALL_PROGRESS,
} from '../../../utils/status/host';
import { getPvStatus, PV_STATUS_ALL_ERROR, PV_STATUS_ALL_PROGRESS } from '../../../utils/status/pv';

import { getCapacityStats } from '../../../selectors';

// same as InventoryItemStatus props
export const STATUS_RESULT_OK = 'ok';
export const STATUS_RESULT_WARN = 'warn';
export const STATUS_RESULT_ERROR = 'error';
export const STATUS_RESULT_IN_PROGRESS = 'inProgress';
export const STATUS_RESULT_OFF = 'off';

const mapStatuses = (entities, mapEntityToStatusResult) => {
  const result = {
    [STATUS_RESULT_OK]: 0,
    [STATUS_RESULT_WARN]: 0,
    [STATUS_RESULT_ERROR]: 0,
    [STATUS_RESULT_IN_PROGRESS]: 0,
    [STATUS_RESULT_OFF]: 0,
    count: null,
  };

  if (entities) {
    entities.forEach(entity => {
      const resolvedStatusResult = (mapEntityToStatusResult && mapEntityToStatusResult(entity)) || STATUS_RESULT_OK;
      result[resolvedStatusResult]++;
    });
    result.count = entities.length;
  }

  return result;
};

const resolveStatusResult = (status, mapper) =>
  Object.keys(mapper).find(key => mapper[key].includes(status)) || STATUS_RESULT_OK;

export const mapNodesToProps = nodes =>
  mapStatuses(nodes, node =>
    resolveStatusResult(getNodeStatus(node).status, {
      [STATUS_RESULT_ERROR]: NODE_STATUS_ALL_ERROR,
      [STATUS_RESULT_WARN]: NODE_STATUS_ALL_WARN,
    })
  );

export const mapVmsToProps = (vms, pods, migrations) =>
  mapStatuses(vms, vm =>
    resolveStatusResult(getVmStatus(vm, pods, migrations).status, {
      [STATUS_RESULT_WARN]: VM_STATUS_ALL_WARNING,
      [STATUS_RESULT_ERROR]: VM_STATUS_ALL_ERROR,
      [STATUS_RESULT_IN_PROGRESS]: VM_STATUS_ALL_PROGRESS,
      [STATUS_RESULT_OFF]: VM_STATUS_ALL_OFF,
    })
  );

export const mapPvcsToProps = pvcs =>
  mapStatuses(pvcs, pvc =>
    resolveStatusResult(getPvcStatus(pvc).status, {
      [STATUS_RESULT_ERROR]: PVC_STATUS_ALL_ERROR,
      [STATUS_RESULT_IN_PROGRESS]: PVC_STATUS_ALL_PROGRESS,
    })
  );

export const mapPvsToProps = pvs =>
  mapStatuses(pvs, pv =>
    resolveStatusResult(getPvStatus(pv).status, {
      [STATUS_RESULT_ERROR]: PV_STATUS_ALL_ERROR,
      [STATUS_RESULT_IN_PROGRESS]: PV_STATUS_ALL_PROGRESS,
    })
  );

export const mapPodsToProps = pods =>
  mapStatuses(pods, pod =>
    resolveStatusResult(getPodStatus(pod).status, {
      [STATUS_RESULT_ERROR]: POD_STATUS_ALL_ERROR,
      [STATUS_RESULT_IN_PROGRESS]: POD_STATUS_ALL_PROGRESS,
    })
  );

export const mapHostsToProps = hosts =>
  mapStatuses(hosts, host =>
    resolveStatusResult(getHostStatus(host).status, {
      [STATUS_RESULT_ERROR]: HOST_STATUS_ALL_ERROR,
      [STATUS_RESULT_WARN]: HOST_STATUS_ALL_WARN,
      [STATUS_RESULT_IN_PROGRESS]: HOST_STATUS_ALL_PROGRESS,
    })
  );

export const mapDiskStatsToProps = (cephOsdUp, cephOsdDown) => {
  const result = {
    [STATUS_RESULT_OK]: 0,
    [STATUS_RESULT_ERROR]: 0,
    count: cephOsdUp && cephOsdDown ? 0 : null,
  };

  const cephOsdUpCount = getCapacityStats(cephOsdUp);
  const cephOsdDownCount = getCapacityStats(cephOsdDown);

  if (cephOsdUpCount || cephOsdDownCount) {
    result[STATUS_RESULT_OK] = cephOsdUpCount;
    result[STATUS_RESULT_ERROR] = cephOsdDownCount;
    result.count = cephOsdUpCount + cephOsdDownCount;
  }

  return result;
};
