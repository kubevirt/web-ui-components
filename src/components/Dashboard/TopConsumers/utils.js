export const getMetricConsumers = ({ consumers }) => {
  const max = Math.max(...consumers.map(c => c.usage));
  return consumers
    .map(c => ({
      now: Math.round((100 * c.usage) / max),
      description: c.name,
      label: c.label,
    }))
    .sort((a, b) => b.now - a.now);
};
