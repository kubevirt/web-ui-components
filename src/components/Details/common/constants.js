export const DESCRIPTION_KEY = 'description';
export const FLAVOR_KEY = 'flavor';
