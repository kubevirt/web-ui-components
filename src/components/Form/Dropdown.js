import React from 'react';
import PropTypes from 'prop-types';
import { ButtonGroup, DropdownButton, MenuItem, noop } from 'patternfly-react';

export const Dropdown = ({ id, value, disabled, onChange, onBlur, choices }) => (
  <ButtonGroup justified key={id}>
    <DropdownButton
      id={id}
      bsStyle="default"
      className="kubevirt-dropdown"
      title={value}
      onSelect={onChange}
      onBlur={onBlur}
      disabled={disabled}
    >
      {choices.map(choice => {
        const isObject = typeof choice === 'object';
        const key = isObject ? choice.id || choice.name : choice;
        const val = isObject ? choice.name : choice;

        return (
          <MenuItem key={key} eventKey={key}>
            {val}
          </MenuItem>
        );
      })}
    </DropdownButton>
  </ButtonGroup>
);

Dropdown.defaultProps = {
  onChange: noop,
  onBlur: noop,
  disabled: false,
};

Dropdown.propTypes = {
  id: PropTypes.string.isRequired,
  value: PropTypes.string.isRequired,
  choices: PropTypes.array.isRequired,
  onChange: PropTypes.func,
  onBlur: PropTypes.func,
  disabled: PropTypes.bool,
};
