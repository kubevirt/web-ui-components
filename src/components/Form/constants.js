export const TEXT = 'text';
export const TEXT_AREA = 'textarea';
export const DROPDOWN = 'dropdown';
export const CHECKBOX = 'checkbox';
export const POSITIVE_NUMBER = 'positive-number';
export const LABEL = 'label';
export const CUSTOM = 'custom';
