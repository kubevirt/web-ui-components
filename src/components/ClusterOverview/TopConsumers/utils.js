import { formatBytes } from '../../../utils';
import { getConsumers } from '../../../selectors';

export const formatCpu = (cpuTime, fixed = 0) => Number(cpuTime).toFixed(fixed);

export const formatBytesWithUnits = bytes => {
  const formattedBytes = formatBytes(bytes);
  return `${formattedBytes.value} ${formattedBytes.unit}`;
};

export const getMetric = (results, metricName, format) => ({
  isLoading: !results,
  consumers: getConsumers(results, metricName, format),
});
