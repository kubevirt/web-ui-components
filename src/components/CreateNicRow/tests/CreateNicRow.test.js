import React from 'react';
import { cloneDeep } from 'lodash';
import { shallow, mount } from 'enzyme';
import { MenuItem } from 'patternfly-react';

import { CreateNicRow } from '..';

import { Dropdown } from '../../Form/Dropdown';
import CreateNicRowFixture from '../fixtures/CreateNicRow.fixture';
import { networkConfigs } from '../../../k8s/mock_network';
import { cloudInitTestVm } from '../../../k8s/mock_vm/cloudInitTestVm.mock';
import { getName } from '../../../utils';
import { Loading } from '../../Loading';
import { POD_NETWORK } from '../../../constants';

const testCreateNicRow = (networks, nic = {}) => (
  <CreateNicRow {...CreateNicRowFixture.props} networks={networks} nic={nic} />
);

const getNetworkConfigs = component =>
  component
    .find('#network-type')
    .find(Dropdown)
    .find(MenuItem);

describe('<CreateNicRow />', () => {
  it('renders correctly', () => {
    const component = shallow(testCreateNicRow([]));
    expect(component).toMatchSnapshot();
  });

  it('filters already used network definitions', () => {
    const nic = {
      vm: cloudInitTestVm,
    };

    const component = mount(testCreateNicRow(networkConfigs, nic));
    expect(getNetworkConfigs(component)).toHaveLength(1);
    expect(
      getNetworkConfigs(component)
        .find('a')
        .text()
    ).toEqual(getName(networkConfigs[0]));

    const vmWithoutPodNetwork = cloneDeep(cloudInitTestVm);

    delete vmWithoutPodNetwork.spec.template.spec.domain.devices.interfaces;
    delete vmWithoutPodNetwork.spec.template.spec.networks;
    vmWithoutPodNetwork.spec.template.spec.domain.devices.autoAttachPodInterface = false;

    const nicNoPod = {
      vm: vmWithoutPodNetwork,
    };

    component.setProps({
      nic: nicNoPod,
    });
    expect(getNetworkConfigs(component)).toHaveLength(2);
    expect(
      getNetworkConfigs(component)
        .at(0)
        .find('a')
        .text()
    ).toEqual(getName(networkConfigs[0]));
    expect(
      getNetworkConfigs(component)
        .at(1)
        .find('a')
        .text()
    ).toEqual(POD_NETWORK);

    const vmWithMultusNetwork = cloneDeep(vmWithoutPodNetwork);

    vmWithMultusNetwork.spec.template.spec.domain.devices.interfaces = [
      {
        name: 'fooNetwork',
        bridge: {},
      },
    ];
    vmWithMultusNetwork.spec.template.spec.networks = [
      {
        name: 'fooNetwork',
        multus: {
          networkName: getName(networkConfigs[0]),
        },
      },
    ];

    const nicWithMultus = {
      vm: vmWithMultusNetwork,
    };

    component.setProps({
      nic: nicWithMultus,
    });
    expect(getNetworkConfigs(component)).toHaveLength(1);
    expect(
      getNetworkConfigs(component)
        .find('a')
        .text()
    ).toEqual(POD_NETWORK);
  });

  it('shows loading while getting network configs', () => {
    const component = mount(testCreateNicRow(undefined));
    expect(component.find(Loading).exists()).toBeTruthy();

    component.setProps({
      networks: [],
    });

    expect(component.find(Loading).exists()).toBeFalsy();
  });

  it('shows loading while patching vm', () => {
    const nic = {
      vm: cloudInitTestVm,
      creating: true,
    };

    const component = mount(testCreateNicRow(networkConfigs, nic));
    expect(component.find(Loading).exists()).toBeTruthy();

    nic.creating = false;
    component.setProps({
      nic,
    });
    expect(component.find(Loading).exists()).toBeFalsy();
  });
});
