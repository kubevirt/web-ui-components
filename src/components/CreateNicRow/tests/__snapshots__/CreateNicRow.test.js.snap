// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateNicRow /> renders correctly 1`] = `
<CreateDeviceRow
  LoadingComponent={[Function]}
  columnSizes={
    Array [
      Object {
        "lg": 3,
        "md": 3,
        "sm": 3,
        "xs": 3,
      },
      Object {
        "lg": 3,
        "md": 3,
        "sm": 3,
        "xs": 3,
      },
      Object {
        "lg": 2,
        "md": 2,
        "sm": 2,
        "xs": 2,
      },
      Object {
        "lg": 2,
        "md": 2,
        "sm": 2,
        "xs": 2,
      },
      Object {
        "lg": 2,
        "md": 2,
        "sm": 2,
        "xs": 2,
      },
    ]
  }
  device={Object {}}
  deviceFields={
    Object {
      "binding": Object {
        "choices": Array [
          "masquerade",
          "bridge",
          "sriov",
        ],
        "defaultValue": "--- Select binding ---",
        "disabled": undefined,
        "id": "binding",
        "required": true,
        "title": "Binding method",
        "type": "dropdown",
      },
      "mac": Object {
        "disabled": false,
        "id": "mac-address",
        "title": "MAC Address",
      },
      "model": Object {
        "id": "nic-model",
        "type": "label",
      },
      "name": Object {
        "disabled": undefined,
        "id": "nic-name",
        "required": true,
        "title": "Name",
        "validate": [Function],
      },
      "network": Object {
        "choices": Array [
          Object {
            "name": "Pod Networking",
            "networkType": "pod",
          },
        ],
        "defaultValue": "--- Select Network Definition ---",
        "disabled": false,
        "id": "network-type",
        "required": true,
        "title": "Network",
        "type": "dropdown",
      },
    }
  }
  onAccept={[Function]}
  onCancel={[Function]}
  onChange={[Function]}
/>
`;
